# 제로베이스 데이터 사이언스 스쿨 3기 SQL 과제 2
- 유가 데이터 분석

### 8문제 총 100점

- 1번 5점
- 2번 5점
- 3번 10점
- 4번 30점
- 5번 10점
- 6번 10점
- 7번 15점
- 8번 15점

### 1 ~ 8번 모두 본 노트북 파일에 답안 작성해서 제출해주세요 :)

---

In [124]:
import mysql.connector
import pandas as pd
import time
from selenium import webdriver
from selenium.webdriver.common.by import By
from bs4 import BeautifulSoup
import re
import numpy as np
from tqdm import tqdm
import googlemaps

문제 1.

주유소 데이터를 저장할 테이블을 다음의 구조로 생성하세요. (PDF 파일 참고)

```
create table GAS_BRAND(
    id int not null auto_increment,
    name varchar(16),
    primary key (id)
);

create table GAS_STATION(
    id int not null auto_increment,
    brand int,
    name varchar(32),
    city char(2),
    gu varchar(10),
    address varchar(128),
    gasoline int,
    diesel 
    self BOOL,int,
    car_wash BOOL,
    charging_station BOOL,
    car_maintenance BOOL,
    convenience_store BOOL,
    24_hour BOOL,
    lat DECIMAL(16,14),
    lng decimal(17,14),
    primary key (id),
    constraint oil foreign key (brand) references GAS_BRAND(id)
);
```

문제 2.

Python 코드로 GAS_BRAND 데이터를 다음과 같이 입력하고 확인하세요. (PDF 파일 참고)

In [70]:
remote = mysql.connector.connect(
    host = 'database-1.cfc20ugqgb4j.ap-northeast-2.rds.amazonaws.com',
    port = 3306,
    user = 'oneday',
    password = '1234',
    database = 'oneday'
)

In [74]:
cur = remote.cursor()
sql = open('oil3.sql', encoding='UTF8').read()

In [75]:
for  result_iterator in cur.execute(sql, multi=True):
    if result_iterator.with_rows:
        print(result_iterator.fetchall())
    else:
        print(result_iterator.statement)

insert into GAS_BRAND VALUES (1, 'SK에너지')
insert into GAS_BRAND VALUES (2, '현대오일뱅크')
insert into GAS_BRAND VALUES (3, 'GS칼텍스')
insert into GAS_BRAND VALUES (4, 'S-OIL')
insert into GAS_BRAND VALUES (5, '알뜰주유소')
insert into GAS_BRAND VALUES (6, '자가상표')


In [76]:
remote.commit()
remote.close()

제출 1.
- Table 생성 결과 : Desc GAS_BRAND; Desc GAS_STATION;

```
+-------+-------------+------+-----+---------+----------------+
| Field | Type        | Null | Key | Default | Extra          |
+-------+-------------+------+-----+---------+----------------+
| id    | int         | NO   | PRI | NULL    | auto_increment |
| name  | varchar(16) | YES  |     | NULL    |                |
+-------+-------------+------+-----+---------+----------------+

+-------------------+----------------+------+-----+---------+----------------+
| Field             | Type           | Null | Key | Default | Extra          |
+-------------------+----------------+------+-----+---------+----------------+
| id                | int            | NO   | PRI | NULL    | auto_increment |
| brand             | int            | YES  | MUL | NULL    |                |
| name              | varchar(32)    | YES  |     | NULL    |                |
| city              | char(2)        | YES  |     | NULL    |                |
| gu                | varchar(10)    | YES  |     | NULL    |                |
| address           | varchar(128)   | YES  |     | NULL    |                |
| gasoline          | int            | YES  |     | NULL    |                |
| diesel            | int            | YES  |     | NULL    |                |
| car_wash          | tinyint(1)     | YES  |     | NULL    |                |
| charging_station  | tinyint(1)     | YES  |     | NULL    |                |
| car_maintenance   | tinyint(1)     | YES  |     | NULL    |                |
| convenience_store | tinyint(1)     | YES  |     | NULL    |                |
| 24_hour           | tinyint(1)     | YES  |     | NULL    |                |
| lat               | decimal(16,14) | YES  |     | NULL    |                |
| lng               | decimal(17,14) | YES  |     | NULL    |                |
+-------------------+----------------+------+-----+---------+----------------+

```

제출 2.
- GAS_BRAND 조회 결과 : SELECT * FROM GAS_BRAND;

```
+----+--------------+
| id | name         |
+----+--------------+
|  1 | SK에너지     |
|  2 | 현대오일뱅크 |
|  3 | GS칼텍스     |
|  4 | S-OIL        |
|  5 | 알뜰주유소   |
|  6 | 자가상표     |
+----+--------------+
```

문제 3.

다음의 함수와 그 함수를 테스트하는 코드를 작성하세요. (PDF 파일 참고)

- 화폐단위 문자형을 입력받아 숫자형으로 반환하는 함수 (테스트 입력 : ‘1,000’)
- 주유소 브랜드를 입력하면 GAS_BRAND 데이터를 참고하여 ID 를 반환하는 함수 (테스트 입력 : ‘SK에너지’)
- 주소를 입력받아 구 이름을 반환하는 함수 (테스트 입력 : ‘서울시 강남구 헌릉로 730’)
- 주소를 입력받아 위도, 경도를 반환하는 함수 (테스트 입력 : ‘서울시 강남구 헌릉로 730’)

제출 3.
- 각 함수 및 테스트 코드와 실행결과

In [125]:
gmaps_key = 'AIzaSyDS8JfW9b9Fc_DasLH4INDxCoy8aU79UfA'
gmaps = googlemaps.Client(key=gmaps_key)

In [26]:
# 화폐단위 문자형을 입력받아 숫자형으로 반환하는 함수 (테스트 입력 : ‘1,000’)
def remove_commas(number_str):
    # 숫자 문자열에서 ',' 제거
    result = number_str.replace(',', '')
    return int(result)

remove_commas('1,000')

1000

In [53]:
#주유소 브랜드를 입력하면 GAS_BRAND 데이터를 참고하여 ID 를 반환하는 함수 (테스트 입력 : ‘SK에너지’)
def brand_search(n):
    if n == 'SK에너지':
        return 1

    elif n == 'HD현대오일뱅크':
        return 2

    elif n == 'GS칼텍스':
        return 3

    elif n == 'S-OIL':
        return 4

    elif n == '알뜰주유소':
        return 5

    else:
        return 6

brand_search('SK에너지')

1

In [28]:
# 주소를 입력받아 구 이름을 반환하는 함수 (테스트 입력 : ‘서울시 강남구 헌릉로 730’)
def split_adr(n):
    tmp  = re.split(' ', n)
    return tmp[1]
split_adr('서울시 강남구 헌릉로 730')

'강남구'

In [29]:
# 주소를 입력받아 위도, 경도를 반환하는 함수 (테스트 입력 : ‘서울시 강남구 헌릉로 730’)
def get_lat_lng(n):
    tmp = gmaps.geocode(n, language='ko')
    lat = tmp[0].get('geometry')['location']['lat']
    lat = round(lat, 7)
    lng = tmp[0].get('geometry')['location']['lng']
    lng = round(lng, 7)
    return lat, lng

get_lat_lng('서울시 강남구 헌릉로 730')

(37.4665675, 127.1190803)

문제 4.

Python 코드에서 주유소 페이지에서 데이터를 가져올때, GAS_STATION 테이블에 바로 입력하도록 수정하세요. (앞서 생 성한 함수활용)
- 주의. city 는 ‘서울’ 로 고정, 부가정보 데이터 타입
- 입력된 데이터의 총 갯수를 쿼리하여 결과를 확인합니다. - 입력된 데이터 상위 10개를 쿼리하여 결과를 확인합니다.


제출 4.
- 주유소 데이터 관련 코드 (ipynb)



In [32]:
url = "https://www.opinet.co.kr/searRgSelect.do"
driver = webdriver.Chrome(executable_path="../../eda/driver/chromedriver.exe")
driver.get(url)

In [33]:
# 세차장 경정비 편의점 24시간 태그
driver.find_element(By.CSS_SELECTOR, '#CWSH_YN').click()
driver.find_element(By.CSS_SELECTOR, '#MAINT_YN').click()
driver.find_element(By.CSS_SELECTOR, '#CVS_YN').click()
driver.find_element(By.CSS_SELECTOR, '#SEL24_YN').click()

In [34]:
sido_list_raw = driver.find_element(By.ID, "SIDO_NM0")
sido_list = sido_list_raw.find_elements(By.TAG_NAME, "option")

sido_names = []
sido_names = [option.get_attribute("value") for option in sido_list]
sido_names = sido_names[1:]

In [35]:
gu_list_raw = driver.find_element(By.ID, "SIGUNGU_NM0")
gu_list = gu_list_raw.find_elements(By.TAG_NAME, "option")

gu_names = [option.get_attribute("value") for option in gu_list]
gu_names = gu_names[1:]

In [36]:
element = gu_list_raw = driver.find_element(By.ID, "SIGUNGU_NM0")
element.send_keys(gu_names[1])

In [37]:
driver.find_element(By.CSS_SELECTOR, '#body1 > tr:nth-child(1) > td.rlist > a').click()

In [39]:
req = driver.page_source
soup = BeautifulSoup(req, 'html.parser')

In [77]:
# id
n = 1

In [54]:
# 브랜드 명
brand_search(soup.select_one("#poll_div_nm").text)

2

In [40]:
# 주유소명
soup.select_one("#os_nm").text

'재건에너지 재정제2주유소 고속셀프지점'

In [59]:
city = '서울'

In [45]:
# 구
gu = split_adr(soup.select_one("#rd_addr").text)
gu

'강동구'

In [41]:
# 주소
soup.select_one("#rd_addr").text

'서울특별시 강동구  천호대로 1246 (둔촌제2동)'

In [47]:
# 휘발유
remove_commas(soup.select_one("#b027_p").text)

1559

In [46]:
# 경유
remove_commas(soup.select_one("#d047_p").text)

1469

In [63]:
# 셀프 여부
if soup.find('img', {'alt': '셀프주유소', 'class': 'bul'}):
    print(1)
else:
    print(0)

1


In [64]:
# 새차 충전 경정비 편의점 24시 
if "off" in soup.select_one("#cwsh_yn").get('src'):
    print(1)
else:
    print(0)

0


In [55]:
# 위도 경도
get_lat_lng(soup.select_one("#rd_addr").text)

(37.5365133, 127.1493588)

```
mysql bool 데이터 처리 방법
https://medium.com/daangn/mysql-boolean-%EC%BB%AC%EB%9F%BC-7abd9b35c664
```

In [96]:
conn = mysql.connector.connect(
    host = 'database-1.cfc20ugqgb4j.ap-northeast-2.rds.amazonaws.com',
    port = 3306,
    user = 'oneday',
    password = '1234',
    database = 'oneday'
)

In [91]:
cursor = conn.cursor(buffered=True)

In [92]:
sql = 'insert into GAS_STATION values (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)'

In [93]:
# 오류발생시 초기화
n = 1

In [94]:
for gu in gu_names:
    element = gu_list_raw = driver.find_element(By.ID, "SIGUNGU_NM0")
    element.send_keys(gu)
    search_result = int(driver.find_element_by_css_selector("#totCnt").text)
    
    for i in range(1, search_result +1):
        driver.find_element_by_css_selector(f'#body1 > tr:nth-child({i}) > td.rlist > a').click()
        html = driver.page_source
        soup = BeautifulSoup(html,"html.parser")
        
        # 초기화
        datas = []
        # id
        datas.append(n)
        # 브랜드명
        datas.append(brand_search(soup.select_one("#poll_div_nm").text))  
        # 주유소명
        datas.append(soup.select_one("#os_nm").text)
        # 서울
        datas.append(city)
        # 구
        datas.append(split_adr(soup.select_one("#rd_addr").text))
        # 주소
        datas.append(soup.select_one("#rd_addr").text)

        # 휘발유
        datas.append(remove_commas(soup.select_one("#b027_p").text))
        # 경유
        if soup.select_one("#d047_p").text == '':
            datas.append(0)
        else:
            datas.append(remove_commas(soup.select_one("#d047_p").text))
        
        # 셀프여부
        if soup.find('img', {'alt': '셀프주유소', 'class': 'bul'}):
            datas.append(1)
        else:
            datas.append(0)
        # 세차장
        if "off" in soup.select_one("#cwsh_yn").get('src'):
            datas.append(0)
        else:
            datas.append(1)
        # 충전소
        if "off" in soup.select_one("#lpg_yn").get('src'):
            datas.append(0)
        else:
            datas.append(1)
        # 경정비
        if "off" in soup.select_one("#maint_yn").get('src'):
            datas.append(0)
        else:
            datas.append(1)
        # 편의점
        if "off" in soup.select_one("#cvs_yn").get('src'):
            datas.append(0)
        else:
            datas.append(1)
        # 24시간
        if "off" in soup.select_one("#sel24_yn").get('src'):
            datas.append(0)
        else:
            datas.append(1)

        # 위도 경도
        tmp = get_lat_lng(soup.select_one("#rd_addr").text)
        lat = tmp[0]
        lng = tmp[1]
        datas.append(lat)
        datas.append(lng)
        cursor.execute(sql, tuple(datas))
        print(tuple(datas))
        conn.commit() 
        time.sleep(0.5)
        n += 1

(1, 1, '(주)보성 세곡주유소', '서울', '강남구', '서울 강남구  헌릉로 731 (세곡동)', 1559, 1469, 1, 1, 0, 0, 0, 0, 37.4671565, 127.1187081)
(2, 2, 'HD현대오일뱅크㈜직영 산성셀프주유소', '서울', '강남구', '서울 강남구 헌릉로 730', 1574, 1494, 1, 1, 0, 0, 0, 0, 37.4665675, 127.1190803)
(3, 1, '자곡셀프주유소', '서울', '강남구', '서울 강남구 밤고개로 120 (자곡동)', 1598, 1559, 1, 1, 0, 0, 0, 0, 37.4834002, 127.1040458)
(4, 4, '극동유화㈜ 개나리주유소', '서울', '강남구', '서울 강남구 언주로 423 (역삼동)', 1615, 1535, 1, 1, 0, 1, 0, 1, 37.5000776, 127.0385419)
(5, 3, '방죽주유소', '서울', '강남구', '서울 강남구 밤고개로 215 (율현동)', 1617, 1539, 1, 1, 0, 0, 0, 0, 37.4749926, 127.1068866)
(6, 2, 'HD현대오일뱅크㈜직영 도곡셀프주유소', '서울', '강남구', '서울 강남구  남부순환로 2718 (도곡2동)', 1619, 1539, 1, 1, 0, 1, 0, 0, 37.4854623, 127.0432181)
(7, 1, '오일프러스 셀프', '서울', '강남구', '서울 강남구 남부순환로 2651 (도곡동)', 1625, 1545, 1, 1, 0, 1, 0, 1, 37.485839, 127.0396721)
(8, 2, '삼성동주유소', '서울', '강남구', '서울 강남구 테헤란로 619 (삼성동)', 1635, 1565, 1, 1, 0, 0, 0, 0, 37.5098576, 127.0651101)
(9, 2, 'HD현대오일뱅크㈜직영 유진주유소', '서울', '강남구', '서울 강남구 논현로 152 (도곡동)', 1639, 1564, 0, 1, 0

In [98]:
driver.quit()

In [104]:
remote.close()

문제 5.

시각화 프로젝트를 위하여 다음의 규칙으로 쿼리하여 CSV 파일로 저장합니다. (Python 코드로 작성)
- 전체 데이터를 가져오는데, 주유소 브랜드 아이디 대신 브랜드명이 표시되어야 합니다. (정렬 : 주유소 매장 아이디 순)
- 다음의 형식으로 저장되어야 함 (브랜드 이름, 칼럼 명 주의, id : GAS_STORE.id)

제출 5.
- 시각화 프로젝트 관련 코드 (ipynb), 결과 파일 (csv)

In [105]:
conn = mysql.connector.connect(
    host = 'database-1.cfc20ugqgb4j.ap-northeast-2.rds.amazonaws.com',
    port = 3306,
    user = 'oneday',
    password = '1234',
    database = 'oneday'
)
cursor = conn.cursor(buffered=True)

In [107]:
sql = 'select * from GAS_STATION'

In [108]:
cursor.execute(sql)

result = cursor.fetchall()
for row in result:
    print(row)

(1, 1, '(주)보성 세곡주유소', '서울', '강남구', '서울 강남구  헌릉로 731 (세곡동)', 1559, 1469, 1, 1, 0, 0, 0, 0, Decimal('37.46715650000000'), Decimal('127.11870810000000'))
(2, 2, 'HD현대오일뱅크㈜직영 산성셀프주유소', '서울', '강남구', '서울 강남구 헌릉로 730', 1574, 1494, 1, 1, 0, 0, 0, 0, Decimal('37.46656750000000'), Decimal('127.11908030000000'))
(3, 1, '자곡셀프주유소', '서울', '강남구', '서울 강남구 밤고개로 120 (자곡동)', 1598, 1559, 1, 1, 0, 0, 0, 0, Decimal('37.48340020000000'), Decimal('127.10404580000000'))
(4, 4, '극동유화㈜ 개나리주유소', '서울', '강남구', '서울 강남구 언주로 423 (역삼동)', 1615, 1535, 1, 1, 0, 1, 0, 1, Decimal('37.50007760000000'), Decimal('127.03854190000000'))
(5, 3, '방죽주유소', '서울', '강남구', '서울 강남구 밤고개로 215 (율현동)', 1617, 1539, 1, 1, 0, 0, 0, 0, Decimal('37.47499260000000'), Decimal('127.10688660000000'))
(6, 2, 'HD현대오일뱅크㈜직영 도곡셀프주유소', '서울', '강남구', '서울 강남구  남부순환로 2718 (도곡2동)', 1619, 1539, 1, 1, 0, 1, 0, 0, Decimal('37.48546230000000'), Decimal('127.04321810000000'))
(7, 1, '오일프러스 셀프', '서울', '강남구', '서울 강남구 남부순환로 2651 (도곡동)', 1625, 1545, 1, 1, 0, 1, 0, 1, De

In [121]:
df = pd.DataFrame(result)
df.columns = ['id', 'brand', 'name', 'city', 'gu', 'address', 'gasoline', 'diesel', 'self', 'car_wash', 'charging_station', 'car_maintenance' , 'convenience_store',  '24_hour', 'lat', 'lng']
df = df.set_index(keys='id')
df.head()

brand                 name city   gu                address  gasoline  \
id                                                                          
1       1          (주)보성 세곡주유소   서울  강남구  서울 강남구  헌릉로 731 (세곡동)      1559   
2       2  HD현대오일뱅크㈜직영 산성셀프주유소   서울  강남구         서울 강남구 헌릉로 730      1574   
3       1              자곡셀프주유소   서울  강남구  서울 강남구 밤고개로 120 (자곡동)      1598   
4       4         극동유화㈜ 개나리주유소   서울  강남구   서울 강남구 언주로 423 (역삼동)      1615   
5       3                방죽주유소   서울  강남구  서울 강남구 밤고개로 215 (율현동)      1617   

    diesel  self  car_wash  charging_station  car_maintenance  \
id                                                              
1     1469     1         1                 0                0   
2     1494     1         1                 0                0   
3     1559     1         1                 0                0   
4     1535     1         1                 0                1   
5     1539     1         1                 0                0   

    convenience_store  24_hour                lat                 lng  
id                                                                     
1                   0        0  37.46715650000000  127.11870810000000  
2                   0        0  37.46656750000000  127.11908030000000  
3                   0        0  37.48340020000000  127.10404580000000  
4                   0        1  37.50007760000000  127.03854190000000  
5                   0        0  37.47499260000000  127.10688660000000

In [122]:
df['brand'] = df['brand'].replace(1, 'SK에너지')
df['brand'] = df['brand'].replace(2, '현대오일뱅크')
df['brand'] = df['brand'].replace(3, 'GS칼텍스')
df['brand'] = df['brand'].replace(4, 'S-OIL')
df['brand'] = df['brand'].replace(5, '알뜰주유소')
df['brand'] = df['brand'].replace(6, '자가상표')
df.head()

brand                 name city   gu                address  gasoline  \
id                                                                           
1    SK에너지          (주)보성 세곡주유소   서울  강남구  서울 강남구  헌릉로 731 (세곡동)      1559   
2   현대오일뱅크  HD현대오일뱅크㈜직영 산성셀프주유소   서울  강남구         서울 강남구 헌릉로 730      1574   
3    SK에너지              자곡셀프주유소   서울  강남구  서울 강남구 밤고개로 120 (자곡동)      1598   
4    S-OIL         극동유화㈜ 개나리주유소   서울  강남구   서울 강남구 언주로 423 (역삼동)      1615   
5    GS칼텍스                방죽주유소   서울  강남구  서울 강남구 밤고개로 215 (율현동)      1617   

    diesel  self  car_wash  charging_station  car_maintenance  \
id                                                              
1     1469     1         1                 0                0   
2     1494     1         1                 0                0   
3     1559     1         1                 0                0   
4     1535     1         1                 0                1   
5     1539     1         1                 0                0   

    convenience_store  24_hour                lat                 lng  
id                                                                     
1                   0        0  37.46715650000000  127.11870810000000  
2                   0        0  37.46656750000000  127.11908030000000  
3                   0        0  37.48340020000000  127.10404580000000  
4                   0        1  37.50007760000000  127.03854190000000  
5                   0        0  37.47499260000000  127.10688660000000

In [123]:
df.to_csv('oil.csv')

문제 6.

위도, 경도 정보를 이용하여 미왕빌딩에서 1킬로 이내에 위치한 주유소 정보를 검색하세요.

- 주유소 아이디, 주유소 브랜드명, 주유소 매장명, 주소, 미왕빌딩으로 부터의 거리 (km)

제출 6.
- 쿼리, 결과

https://yusang.tistory.com/48

In [130]:
tmp = get_lat_lng('서울특별시 강남구 강남대로 364')
lat = tmp[0]
lng = tmp[1]
lat, lng

(37.4955525, 127.0292924)

SELECT
    id, brand, name, address,
	 ROUND((6371*acos(cos(radians(37.4955525))*cos(radians(lat))*cos(radians(lng)
    -radians(127.0292924))+sin(radians(37.4955525))*sin(radians(lat)))), 2)
    AS distance

FROM GAS_STATION
	HAVING distance < 1

ORDER BY distance;

``` distance;

+-----+-------+-------------------------------------+-----------------------------------+----------+
| id  | brand | name                                | address                           | distance |
+-----+-------+-------------------------------------+-----------------------------------+----------+
|  12 |     1 | SK서광주유소                        | 서울 강남구 역삼로 142            |     0.53 |
| 210 |     2 | HD현대오일뱅크㈜직영 서초제일주유소 | 서울 서초구 사임당로 116 (서초동) |     0.81 |
|   4 |     4 | 극동유화㈜ 개나리주유소             | 서울 강남구 언주로 423 (역삼동)   |     0.96 |
+-----+-------+-------------------------------------+-----------------------------------+----------+
```

문제 7.

위도, 경도 정보를 이용하여 미왕빌딩에서 셀프주유가 가능하고 24시간이면서 편의점이 있는 가장 가까운 주유소 10개를 휘 발유 가격이 가장 저렴한 순으로 정렬하여 조회하세요.

- 주유소 아이디, 주유소 브랜드명, 주유소 매장명, 주소, 휘발유 가격, 부가정보 (셀프, 24시간, 편의점 여부), 미왕빌딩으로 부터의 거리 (km)


제출 7.
- 쿼리, 결과

SELECT
    id, brand, name, address, gasoline, self, 24_hour, convenience_store,
	 ROUND((6371*acos(cos(radians(37.4955525))*cos(radians(lat))*cos(radians(lng)
    -radians(127.0292924))+sin(radians(37.4955525))*sin(radians(lat)))), 2)
    AS distance

FROM GAS_STATION
where self = 1 and 24_hour = 1 and convenience_store = 1
ORDER BY distance, gasoline
limit 10;

```
+-----+-------+-----------------------------------------+-----------------------------------+----------+------+---------+-------------------+----------+
| id  | brand | name                                    | address                           | gasoline | self | 24_hour | convenience_store | distance |
+-----+-------+-----------------------------------------+-----------------------------------+----------+------+---------+-------------------+----------+
| 215 |     1 | 대신석유(주)대우주유소                  | 서울 서초구 양재대로 173 (양재동) |     1649 |    1 |       1 |                 1 |     3.71 |
| 196 |     6 | 만남의광장주유소                        | 서울 서초구 양재대로12길 73-71    |     1514 |    1 |       1 |                 1 |      4.1 |
| 253 |     3 | 지에스칼텍스(주)가든파이브주유소        | 서울 송파구 탄천동로 740          |     1563 |    1 |       1 |                 1 |     5.29 |
| 262 |     1 | 백제고분로주유소                        | 서울 송파구 오금로 143 (방이동)   |     1583 |    1 |       1 |                 1 |     7.95 |
| 190 |     1 | 행촌제2주유소                           | 서울 서대문구 성산로 490 (대신동) |     1595 |    1 |       1 |                 1 |     10.7 |
| 303 |     1 | 행촌주유소                              | 서울 영등포구 도신로 130 (신길동) |     1499 |    1 |       1 |                 1 |    10.95 |
| 312 |     1 | SK에너지(주) 기린주유소                 | 서울 영등포구 선유로 270          |     1599 |    1 |       1 |                 1 |    12.32 |
| 121 |     3 | 대득에너지(금천셀프주유소)              | 서울 금천구  시흥대로 81 (시흥동) |     1575 |    1 |       1 |                 1 |    12.68 |
| 183 |     2 | (주)가재울뉴타운주유소                  | 서울 서대문구 모래내로 205        |     1523 |    1 |       1 |                 1 |    12.79 |
| 288 |     2 | 현대오일뱅크(주)직영 양천현대셀프주유소 | 서울 양천구 안양천로 1179 (목동)  |     1579 |    1 |       1 |                 1 |     14.2 |
+-----+-------+-----------------------------------------+-----------------------------------+----------+------+---------+-------------------+----------+
```

문제 8.

구별로 주유소 브랜드 별 휘발유 평균가격을 조회하여 저렴한 순으로 출력하세요.

- 구 이름, 주유소 브랜드 이름, 휘발유 평균 가격

제출 8.
- 쿼리, 결과

SELECT s.gu, b.name, avg(s.gasoline)
    -> FROM GAS_BRAND as b ,GAS_STATION as s
    -> GROUP BY s.gu, b.name
    -> order by avg(gasoline);

```
+----------+--------------+-----------------+
| gu       | name         | avg(s.gasoline) |
+----------+--------------+-----------------+
| 도봉구   | 자가상표     |       1539.4286 |
| 도봉구   | 알뜰주유소   |       1539.4286 |
| 도봉구   | S-OIL        |       1539.4286 |
| 도봉구   | GS칼텍스     |       1539.4286 |
| 도봉구   | 현대오일뱅크 |       1539.4286 |
| 도봉구   | SK에너지     |       1539.4286 |
| 은평구   | 자가상표     |       1542.3333 |
| 은평구   | 알뜰주유소   |       1542.3333 |
| 은평구   | S-OIL        |       1542.3333 |
| 은평구   | GS칼텍스     |       1542.3333 |
| 은평구   | 현대오일뱅크 |       1542.3333 |
| 은평구   | SK에너지     |       1542.3333 |
| 중랑구   | 자가상표     |       1543.2727 |
| 중랑구   | 알뜰주유소   |       1543.2727 |
| 중랑구   | S-OIL        |       1543.2727 |
| 중랑구   | GS칼텍스     |       1543.2727 |
| 중랑구   | 현대오일뱅크 |       1543.2727 |
| 중랑구   | SK에너지     |       1543.2727 |
| 강북구   | 자가상표     |       1545.9000 |
| 강북구   | 알뜰주유소   |       1545.9000 |
| 강북구   | S-OIL        |       1545.9000 |
| 강북구   | GS칼텍스     |       1545.9000 |
| 강북구   | 현대오일뱅크 |       1545.9000 |
| 강북구   | SK에너지     |       1545.9000 |
| 성북구   | 자가상표     |       1553.6250 |
| 성북구   | 알뜰주유소   |       1553.6250 |
| 성북구   | S-OIL        |       1553.6250 |
| 성북구   | GS칼텍스     |       1553.6250 |
| 성북구   | 현대오일뱅크 |       1553.6250 |
| 성북구   | SK에너지     |       1553.6250 |
| 양천구   | 자가상표     |       1564.8500 |
| 양천구   | 알뜰주유소   |       1564.8500 |
| 양천구   | S-OIL        |       1564.8500 |
| 양천구   | GS칼텍스     |       1564.8500 |
| 양천구   | 현대오일뱅크 |       1564.8500 |
| 양천구   | SK에너지     |       1564.8500 |
| 금천구   | 자가상표     |       1570.6250 |
| 금천구   | 알뜰주유소   |       1570.6250 |
| 금천구   | S-OIL        |       1570.6250 |
| 금천구   | GS칼텍스     |       1570.6250 |
| 금천구   | 현대오일뱅크 |       1570.6250 |
| 금천구   | SK에너지     |       1570.6250 |
| 서대문구 | 자가상표     |       1571.7692 |
| 서대문구 | 알뜰주유소   |       1571.7692 |
| 서대문구 | S-OIL        |       1571.7692 |
| 서대문구 | GS칼텍스     |       1571.7692 |
| 서대문구 | 현대오일뱅크 |       1571.7692 |
| 서대문구 | SK에너지     |       1571.7692 |
| 동대문구 | 자가상표     |       1571.9286 |
| 동대문구 | 알뜰주유소   |       1571.9286 |
| 동대문구 | S-OIL        |       1571.9286 |
| 동대문구 | GS칼텍스     |       1571.9286 |
| 동대문구 | 현대오일뱅크 |       1571.9286 |
| 동대문구 | SK에너지     |       1571.9286 |
| 노원구   | 자가상표     |       1573.6154 |
| 노원구   | 알뜰주유소   |       1573.6154 |
| 노원구   | S-OIL        |       1573.6154 |
| 노원구   | GS칼텍스     |       1573.6154 |
| 노원구   | 현대오일뱅크 |       1573.6154 |
| 노원구   | SK에너지     |       1573.6154 |
| 구로구   | 자가상표     |       1577.8125 |
| 구로구   | 알뜰주유소   |       1577.8125 |
| 구로구   | S-OIL        |       1577.8125 |
| 구로구   | GS칼텍스     |       1577.8125 |
| 구로구   | 현대오일뱅크 |       1577.8125 |
| 구로구   | SK에너지     |       1577.8125 |
| 강서구   | 자가상표     |       1595.6667 |
| 강서구   | 알뜰주유소   |       1595.6667 |
| 강서구   | S-OIL        |       1595.6667 |
| 강서구   | GS칼텍스     |       1595.6667 |
| 강서구   | 현대오일뱅크 |       1595.6667 |
| 강서구   | SK에너지     |       1595.6667 |
| 광진구   | 자가상표     |       1595.7273 |
| 광진구   | 알뜰주유소   |       1595.7273 |
| 광진구   | S-OIL        |       1595.7273 |
| 광진구   | GS칼텍스     |       1595.7273 |
| 광진구   | 현대오일뱅크 |       1595.7273 |
| 광진구   | SK에너지     |       1595.7273 |
| 송파구   | 자가상표     |       1597.3333 |
| 송파구   | 알뜰주유소   |       1597.3333 |
| 송파구   | S-OIL        |       1597.3333 |
| 송파구   | GS칼텍스     |       1597.3333 |
| 송파구   | 현대오일뱅크 |       1597.3333 |
| 송파구   | SK에너지     |       1597.3333 |
| 마포구   | 자가상표     |       1600.0000 |
| 마포구   | 알뜰주유소   |       1600.0000 |
| 마포구   | S-OIL        |       1600.0000 |
| 마포구   | GS칼텍스     |       1600.0000 |
| 마포구   | 현대오일뱅크 |       1600.0000 |
| 마포구   | SK에너지     |       1600.0000 |
| 동작구   | 자가상표     |       1601.2857 |
| 동작구   | 알뜰주유소   |       1601.2857 |
| 동작구   | S-OIL        |       1601.2857 |
| 동작구   | GS칼텍스     |       1601.2857 |
| 동작구   | 현대오일뱅크 |       1601.2857 |
| 동작구   | SK에너지     |       1601.2857 |
| 관악구   | 자가상표     |       1605.0000 |
| 관악구   | 알뜰주유소   |       1605.0000 |
| 관악구   | S-OIL        |       1605.0000 |
| 관악구   | GS칼텍스     |       1605.0000 |
| 관악구   | 현대오일뱅크 |       1605.0000 |
| 관악구   | SK에너지     |       1605.0000 |
| 영등포구 | 자가상표     |       1636.9130 |
| 영등포구 | 알뜰주유소   |       1636.9130 |
| 영등포구 | S-OIL        |       1636.9130 |
| 영등포구 | GS칼텍스     |       1636.9130 |
| 영등포구 | 현대오일뱅크 |       1636.9130 |
| 영등포구 | SK에너지     |       1636.9130 |
| 서초구   | 자가상표     |       1665.1200 |
| 서초구   | 알뜰주유소   |       1665.1200 |
| 서초구   | S-OIL        |       1665.1200 |
| 서초구   | GS칼텍스     |       1665.1200 |
| 서초구   | 현대오일뱅크 |       1665.1200 |
| 서초구   | SK에너지     |       1665.1200 |
| 성동구   | 자가상표     |       1714.0909 |
| 성동구   | 알뜰주유소   |       1714.0909 |
| 성동구   | S-OIL        |       1714.0909 |
| 성동구   | GS칼텍스     |       1714.0909 |
| 성동구   | 현대오일뱅크 |       1714.0909 |
| 성동구   | SK에너지     |       1714.0909 |
| 강동구   | 자가상표     |       1714.5000 |
| 강동구   | 알뜰주유소   |       1714.5000 |
| 강동구   | S-OIL        |       1714.5000 |
| 강동구   | GS칼텍스     |       1714.5000 |
| 강동구   | 현대오일뱅크 |       1714.5000 |
| 강동구   | SK에너지     |       1714.5000 |
| 종로구   | 자가상표     |       1804.0000 |
| 종로구   | 알뜰주유소   |       1804.0000 |
| 종로구   | S-OIL        |       1804.0000 |
| 종로구   | GS칼텍스     |       1804.0000 |
| 종로구   | 현대오일뱅크 |       1804.0000 |
| 종로구   | SK에너지     |       1804.0000 |
| 강남구   | 자가상표     |       1860.8065 |
| 강남구   | 알뜰주유소   |       1860.8065 |
| 강남구   | S-OIL        |       1860.8065 |
| 강남구   | GS칼텍스     |       1860.8065 |
| 강남구   | 현대오일뱅크 |       1860.8065 |
| 강남구   | SK에너지     |       1860.8065 |
| 중구     | 자가상표     |       1986.8571 |
| 중구     | 알뜰주유소   |       1986.8571 |
| 중구     | S-OIL        |       1986.8571 |
| 중구     | GS칼텍스     |       1986.8571 |
| 중구     | 현대오일뱅크 |       1986.8571 |
| 중구     | SK에너지     |       1986.8571 |
| 용산구   | 자가상표     |       2111.2727 |
| 용산구   | 알뜰주유소   |       2111.2727 |
| 용산구   | S-OIL        |       2111.2727 |
| 용산구   | GS
```칼텍스     |       2111.2727 |
| 용산구   | 현대오일뱅크 |       2111.2727 |
| 용산구   | SK에너지     |       2111.2727 |
+----------+--------------+-----------------+

In [131]:
sql = 'SELECT s.gu, b.name, avg(s.gasoline) FROM GAS_BRAND as b ,GAS_STATION as s GROUP BY s.gu, b.name order by avg(gasoline);'

In [132]:
conn = mysql.connector.connect(
    host = 'database-1.cfc20ugqgb4j.ap-northeast-2.rds.amazonaws.com',
    port = 3306,
    user = 'oneday',
    password = '1234',
    database = 'oneday'
)
cursor = conn.cursor(buffered=True)

In [133]:
cursor.execute(sql)

result = cursor.fetchall()
for row in result:
    print(row)

('도봉구', '자가상표', Decimal('1539.4286'))
('도봉구', '알뜰주유소', Decimal('1539.4286'))
('도봉구', 'S-OIL', Decimal('1539.4286'))
('도봉구', 'GS칼텍스', Decimal('1539.4286'))
('도봉구', '현대오일뱅크', Decimal('1539.4286'))
('도봉구', 'SK에너지', Decimal('1539.4286'))
('은평구', '자가상표', Decimal('1542.3333'))
('은평구', '알뜰주유소', Decimal('1542.3333'))
('은평구', 'S-OIL', Decimal('1542.3333'))
('은평구', 'GS칼텍스', Decimal('1542.3333'))
('은평구', '현대오일뱅크', Decimal('1542.3333'))
('은평구', 'SK에너지', Decimal('1542.3333'))
('중랑구', '자가상표', Decimal('1543.2727'))
('중랑구', '알뜰주유소', Decimal('1543.2727'))
('중랑구', 'S-OIL', Decimal('1543.2727'))
('중랑구', 'GS칼텍스', Decimal('1543.2727'))
('중랑구', '현대오일뱅크', Decimal('1543.2727'))
('중랑구', 'SK에너지', Decimal('1543.2727'))
('강북구', '자가상표', Decimal('1545.9000'))
('강북구', '알뜰주유소', Decimal('1545.9000'))
('강북구', 'S-OIL', Decimal('1545.9000'))
('강북구', 'GS칼텍스', Decimal('1545.9000'))
('강북구', '현대오일뱅크', Decimal('1545.9000'))
('강북구', 'SK에너지', Decimal('1545.9000'))
('성북구', '자가상표', Decimal('1553.6250'))
('성북구', '알뜰주유소', Decimal('

In [134]:
conn.close()

---